Handles firing closeOnClick before another listener

avpeery opened the pull request:
This PR closes #10029. 

Previous bug required double click to open a popup in new area when using the same popup as the closeOnClick event listener fired after opening the popup event listener. This change moves the closeOnClick listener to be fired first in the case that there are multiple listeners for an event. 

You can manually test this with debug/popup.html and using this code to create the popup: 
`const popup = new mapboxgl.Popup();
popup.setHTML("Hello world");
map.on("click", (e) => popup.setLngLat(e.lngLat).addTo(map));`

**WIP**: this doesn't consider the use-case if a user were to click on the same target intending to close a popup, and would prefer moving into helper function without having the index hard-coded in the logic. 

**Functionality before the change:** 

https://user-images.githubusercontent.com/42715836/129082647-2a0610ba-3daa-494c-87a5-f0dfe175f5d3.mov

**Functionality after the change:**

https://user-images.githubusercontent.com/42715836/129082676-d8b8b2bd-790a-4652-afee-f8eb2a41737c.mov

## Launch Checklist
 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [ ] post benchmark scores
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'

mourner commented on src/util/evented.js:
```diff
@@ -124,6 +124,12 @@ export class Evented {
             const listeners = this._listeners && this._listeners[type] ? this._listeners[type].slice() : [];
+
+            // Check if onClose is an event listener and move to front of list to fire first
+            if (listeners.length > 1 && listeners[1].name === "bound _onClose") {
```
It may work for this specific use case, but this solution is brittle, with possible cases where it breaks and it's very hard to debug — for example, what if a user puts multiple `click` listeners on a popup? Then `_onClose` will not be the second, but later in the array. What if a user adds a listener with the same name? What if a different class in GL JS (not popup) uses the same method name (`_onClose`) and the execution order is swapped unexpectedly?

I think we should find a solution that's more explicit and can't produce unexpected side effects — e.g. firing a `preclick` event before `click` [here](https://github.com/mapbox/mapbox-gl-js/blob/7d2f272e4cf478c286fb115cc7550a19ea477f91/src/ui/handler/map_event.js#L43) and then subscribing to it for `Popup` `closeOnClick`. This is how Leaflet handles it as seen [here](https://github.com/Leaflet/Leaflet/blob/master/src/map/Map.js#L1388-L1397).

avpeery replied:
Thanks so much for the feedback! Yes I wasn't confident on this initial approach. I appreciate sending over the resources and example from leaflet. I implemented a preclick that works manually with the debug file but is breaking some of the unit tests in test/unit/ui/map_events.test.js that I'm looking into now.

mourner reviewed (changes requested):
This looks great to me now, nice work! Let's try adding a unit test for the new behavior, and once it's in it's good to merge.

mourner commented:
Not sure what's going on with the render tests, the changes don't seem to have anything that would cause so many failures so it might be something unrelated to the PR entirely...

mourner commented:
`test-render-prod` completed successfully so it seems like a temporary fluke, triggered a rerun of the failed `render-test` job now.

avpeery commented:
@mourner thanks so much for your help! FYI - it says per the unit tests that the expectation for the default behavior for closeOnClick should be true, but it was set as false (which I just changed in the recent comment), I also added the extra documentation in src/ui/map.js but wondering if that is unneeded and should I remove? Yes, I'm working on a unit test right now!

mourner commented:
> I also added the extra documentation in src/ui/map.js but wondering if that is unneeded and should I remove?

That was a good call, let's keep it — you never know when a user might find something like this useful, and `preclick` is exposed in Leaflet (and has been for years) so there's a positive precedent.

mourner reviewed (approved):
Nice work!
